File: LeetCode/dp/BestTimeToBuyAndSellStock.java
package dp;

public class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) return 0;
        // dp[i][0]代表第i天不持有股票的最大收益
        // dp[i][1]代表第i天持有股票的最大收益
        int n = prices.length;
        int[][] dp = new int[n][2];

        dp[0][0] = 0;
        dp[0][1] = -prices[0];

        for (int i = 1; i < n; i++) {
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
            dp[i][1] = Math.max(dp[i - 1][1], -prices[i]);
        }

        return dp[n - 1][0];
    }

    public static void main(String[] args) {
        BestTimeToBuyAndSellStock solution = new BestTimeToBuyAndSellStock();

        // Test case
        int[] prices = {7, 1, 5, 3, 6, 4};
        System.out.println(solution.maxProfit(prices)); // Output: 5
    }
}

File: LeetCode_Python/greedy/BestTimeToBuyAndSellStock.py
class BestTimeToBuyAndSellStock:
    def maxProfit(self, prices: list[int]) -> int:
        min_price = float('inf')
        max_profit = 0
        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)
        return max_profit


if __name__ == '__main__':
    solution = BestTimeToBuyAndSellStock()
    prices1 = [7, 1, 5, 3, 6, 4]

    print("Expected Output:", 5)
    print("Actual Output:", solution.maxProfit(prices1))

    prices2 = [7, 6, 4, 3, 1]
    print("Expected Output:", 0)
    print("Actual Output:", solution.maxProfit(prices2))
